File: backend/services/cursor-runner/src/otel.ts
/**
 * Env-var-driven OpenTelemetry tracing initialization.
 *
 * Mirrors the workflow-runner pattern (pkg/otel/otel.go): reads
 * OTEL_EXPORTER_OTLP_ENDPOINT and configures an OTLP/gRPC exporter
 * with W3C TraceContext + Baggage propagation. When the env var is unset,
 * initTracing() returns null — zero overhead, no SDK initialization.
 */

export async function initTracing(serviceName: string): Promise<(() => Promise<void>) | null> {
  const endpoint = process.env.OTEL_EXPORTER_OTLP_ENDPOINT;
  if (!endpoint) {
    return null;
  }

  const { NodeTracerProvider } = await import("@opentelemetry/sdk-trace-node");
  const { BatchSpanProcessor } = await import("@opentelemetry/sdk-trace-base");
  const { OTLPTraceExporter } = await import("@opentelemetry/exporter-trace-otlp-grpc");
  const { Resource } = await import("@opentelemetry/resources");
  const { W3CTraceContextPropagator, CompositePropagator } = await import("@opentelemetry/core");
  const { W3CBaggagePropagator } = await import("@opentelemetry/core");
  const otelApi = await import("@opentelemetry/api");

  const resource = new Resource({ "service.name": serviceName });
  const exporter = new OTLPTraceExporter({ url: endpoint });
  const provider = new NodeTracerProvider({ resource });
  provider.addSpanProcessor(new BatchSpanProcessor(exporter));
  provider.register();

  otelApi.propagation.setGlobalPropagator(
    new CompositePropagator({
      propagators: [new W3CTraceContextPropagator(), new W3CBaggagePropagator()],
    }),
  );

  console.log(`OTel tracing enabled (endpoint=${endpoint}, service=${serviceName})`);

  return async () => {
    await provider.shutdown();
  };
}

// Baggage key constants — must match workflow-runner pkg/otel/spans.go.
export const BAGGAGE_EXECUTION_ID = "stigmer.execution_id";
export const BAGGAGE_SESSION_ID = "stigmer.session_id";
export const BAGGAGE_ORG_ID = "stigmer.org_id";

/**
 * Attach key-value pairs to the active OTel context as W3C baggage.
 * No-op when tracing is not initialized. Keys with empty values are skipped.
 */
export async function setBaggage(items: Record<string, string>): Promise<void> {
  try {
    const otelApi = await import("@opentelemetry/api");
    const ctx = otelApi.context.active();
    let bag = otelApi.propagation.getBaggage(ctx) ?? otelApi.createBaggage();
    for (const [k, v] of Object.entries(items)) {
      if (v) {
        bag = bag.setEntry(k, { value: v });
      }
    }
    const newCtx = otelApi.propagation.setBaggage(ctx, bag);
    otelApi.context.with(newCtx, () => {});
  } catch {
    // Tracing not initialized — silently skip.
  }
}

File: client-apps/cli/embedded/agentrunner/source/src/stigmer_runner/worker/otel.py
"""Env-var-driven OpenTelemetry tracing initialization.

Mirrors the workflow-runner pattern (``pkg/otel/otel.go``): reads
``OTEL_EXPORTER_OTLP_ENDPOINT`` and configures an OTLP/gRPC exporter
with W3C TraceContext + Baggage propagation.  When the env var is unset,
``init_tracing`` returns ``None`` — zero overhead, no SDK initialization.
"""

import logging
import os
from collections.abc import Callable

logger = logging.getLogger(__name__)


def init_tracing(service_name: str) -> Callable[[], None] | None:
    """Initialize OTel tracing if ``OTEL_EXPORTER_OTLP_ENDPOINT`` is set.

    Returns a synchronous shutdown callable that flushes buffered spans,
    or ``None`` when tracing is disabled.
    """
    endpoint = os.environ.get("OTEL_EXPORTER_OTLP_ENDPOINT")
    if not endpoint:
        return None

    from opentelemetry import propagate, trace
    from opentelemetry.baggage.propagation import W3CBaggagePropagator
    from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
    from opentelemetry.propagators.composite import CompositePropagator
    from opentelemetry.sdk.resources import Resource
    from opentelemetry.sdk.trace import TracerProvider
    from opentelemetry.sdk.trace.export import BatchSpanProcessor
    from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator

    resource = Resource.create({"service.name": service_name})
    exporter = OTLPSpanExporter(endpoint=endpoint, insecure=True)
    provider = TracerProvider(resource=resource)
    provider.add_span_processor(BatchSpanProcessor(exporter))
    trace.set_tracer_provider(provider)

    propagate.set_global_textmap(
        CompositePropagator([TraceContextTextMapPropagator(), W3CBaggagePropagator()])
    )

    logger.info("OTel tracing enabled (endpoint=%s, service=%s)", endpoint, service_name)

    def shutdown() -> None:
        provider.shutdown()

    return shutdown


# Baggage key constants — must match workflow-runner pkg/otel/spans.go.
BAGGAGE_EXECUTION_ID = "stigmer.execution_id"
BAGGAGE_SESSION_ID = "stigmer.session_id"
BAGGAGE_ORG_ID = "stigmer.org_id"


def init_metrics(service_name: str) -> Callable[[], None] | None:
    """Initialize OTel metrics if ``OTEL_EXPORTER_OTLP_ENDPOINT`` is set.

    Returns a synchronous shutdown callable that flushes buffered metrics,
    or ``None`` when metrics are disabled.
    """
    endpoint = os.environ.get("OTEL_EXPORTER_OTLP_ENDPOINT")
    if not endpoint:
        return None

    from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
    from opentelemetry.sdk.metrics import MeterProvider
    from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
    from opentelemetry.sdk.resources import Resource

    resource = Resource.create({"service.name": service_name})
    exporter = OTLPMetricExporter(endpoint=endpoint, insecure=True)
    reader = PeriodicExportingMetricReader(exporter, export_interval_millis=30_000)
    provider = MeterProvider(resource=resource, metric_readers=[reader])

    from opentelemetry import metrics as otel_metrics
    otel_metrics.set_meter_provider(provider)

    logger.info("OTel metrics enabled (endpoint=%s, service=%s)", endpoint, service_name)

    def shutdown() -> None:
        provider.shutdown()

    return shutdown


def set_baggage(items: dict[str, str]) -> None:
    """Attach key-value pairs to the current context as W3C baggage.

    Must be called inside an async task where ``contextvars`` are active
    (e.g. inside a Temporal activity). Keys with empty values are skipped.
    """
    from opentelemetry import baggage, context as otel_context

    ctx = otel_context.get_current()
    for k, v in items.items():
        if v:
            ctx = baggage.set_baggage(k, v, ctx)
    otel_context.attach(ctx)
